crates/adapters/src/transport/secret_resolver.rs

```diff
@@ -0,0 +1,260 @@
+/// formatted.
+#[derive(Clone, PartialEq, Eq)]
+pub enum SimpleOrSecretString {
+    Simple(String),
+    Secret(String),
+}
+
+impl Display for SimpleOrSecretString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            SimpleOrSecretString::Simple(simple_string) => {
+                write!(f, "Simple({})", simple_string)
+            }
+            SimpleOrSecretString::Secret(_) => {
+                warn!(
+                    "Resolved secret is requested to be formatted. \
+                       It is not allowed to format the internal secret string, \
+                       and as such it is not shown."
+                );
+                write!(f, "Secret(***)")
+            }
+        }
+    }
+}
+
+impl Debug for SimpleOrSecretString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // Use Display formatting
+        write!(f, "{}", self)
+    }
+}
+
+impl SimpleOrSecretString {
+    /// A secret reference can only contain a-z and 0-9. There can be connecting
+    /// hyphens (-), though not at the start, not at the end, and not two in
+    /// succession. It cannot be empty.
+    ///
+    /// Valid examples:
+    /// "example", "example-1", "example-of-this"
+    ///
+    /// Invalid examples:
+    /// "", "example--1", "example_1", "-", "Example#1", "example-"
+    fn is_valid_secret_ref_format(secret_ref: &str) -> bool {
+        let re = Regex::new(r"^[a-z0-9]+(-[a-z0-9]+)*$").unwrap();
+        re.is_match(secret_ref)
+    }
+
+    pub fn new_with_default_volume(value: SimpleStringOrSecretRef) -> Self {
+        Self::new("/etc/secret-volume".to_string(), value)
+    }
+
+    pub fn new(secret_volume_path: String, value: SimpleStringOrSecretRef) -> Self {
+        match value {
+            SimpleStringOrSecretRef::SimpleString(simple_string) => Self::Simple(simple_string),
+            SimpleStringOrSecretRef::SecretRef(secret_ref) => {
+                // Method of resolution
+                // ====================
+                // The only current resolution method is through the mounted secret volume
+                // ("/path/to/secret/volume"). A secret reference "example" will have the
+                // secret string fetched from the file at "/path/to/secret/volume/.example".
+                // Prefixing each secret key with a dot is convention from Kubernetes
+                // documentation: https://kubernetes.io/docs/concepts/configuration/secret/
+                // (such that they are hidden in `ls -l` and require `ls -la` to show)
+
+                // Check format
+                if !Self::is_valid_secret_ref_format(&secret_ref) {
+                    panic!(
+                        "Secret reference does not follow valid format: {}",
+                        secret_ref
+                    );
+                }
+
+                // Create the file path
+                let file_path = secret_volume_path + "/." + secret_ref.as_str();
```

Comment:
For robustness, it's better to construct and join paths using `std::path` primitives.

Reply:
Construction and joining is now done using paths.